torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  auto domain = tv->domain()->hasRFactor() ? tv->domain()->rfactorDomain()
+                                           : tv->domain()->rootDomain();
+
+  TORCH_INTERNAL_ASSERT(domain.size() == indices.size());
+
+  bool is_contiguous = true;
+  for (auto status : tv->domain()->contiguity()) {
+    is_contiguous &= status;
+  }
+
+  // If the tensorview is not contiguous, return inner-most root domain extent
+  if (!is_contiguous) {
+    return domain.back()->extent();
+  }
+
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  // Calculate extent of merged root domains
+  kir::Val* extent = nullptr;
+  for (int i = int(domain.size()) - 1; i >= 0; --i) {
+    auto root_id = domain.at(i);
+    if (root_id->isBroadcast() || root_id->isReduction() ||
+        gpu_lower->trivialReductionInfo().isDerived(root_id)) {
+      continue;
+    } else if (extent == nullptr) {
+      extent = root_id->extent();
+    } else if (extent != nullptr && indices.at(i)->isZeroInt()) {
+      // This root id must be merged and contiguous. Expand the
+      // vectorization partition.
+      extent = ir_builder.mulExpr(extent, root_id->extent());
+    } else {
+      break;
+    }
+  }
+
+  TORCH_INTERNAL_ASSERT(extent != nullptr);
+
+  return extent;
+}
+
+kir::Val* setupNamedScalar(
+    kir::Scope& body,
+    kir::Val* val,
+    const std::string& name,
+    bool address = false) {
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  auto namedScalar = (address) ? ir_builder.namedAddressExpr(name, val)
+                               : ir_builder.namedSetExpr(name, val);
+  auto alloc = ir_builder.create<kir::Allocate>(
+      namedScalar, MemoryType::Local, ir_builder.one());
+  body.push_back(alloc);
+  body.push_back(namedScalar->definition());
+  return namedScalar;
+}
+
+kir::ForLoop* handleMisalignedVectorization(
+    std::vector<kir::ForLoop*> loop_structure,
+    const kir::ForLoop* for_loop) {
+  // for_loop body contains allocate, read, compute, write operations
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+
+  // create new base For-Loop
+  const auto new_loop = ir_builder.create<kir::ForLoop>(
+      for_loop->index(), for_loop->iter_domain(), for_loop->extent());
+
+  // Find child For-Loops and add remaining expressions to base For-Loop
+  auto child_loops = parseVectorizedForLoop(for_loop, new_loop);
+
+  // Find the first vectorize set - either read or write
+  auto vec_expr = findVectorizedSet(loop_structure, child_loops);
+  TORCH_INTERNAL_ASSERT(vec_expr != nullptr);
+  TORCH_INTERNAL_ASSERT(vec_expr->outputs().front()->isA<kir::TensorView>());
+  TORCH_INTERNAL_ASSERT(vec_expr->inputs().front()->isA<kir::TensorView>());
+
+  auto out_tv = vec_expr->outputs().front()->as<kir::TensorView>();
+  auto in_tv = vec_expr->inputs().front()->as<kir::TensorView>();
+
+  // It is assumed that either of input and output is on global memory
+  TORCH_INTERNAL_ASSERT(
```

Comment:
This seems overly redundant, one has to be global and the other has to be local, can't we check for that in one assert?

Reply:
This should be an interface check, though could also be asserted here.

Reply:
@naoyam Couldn't we merge these assertions into a single check?

Reply:
Please feel free to make it simpler. Agree I was overly redundant. I wanted document the assumptions explicitly.

However, please do not remove the assertions from this function entirely. I know they are already validated, but I believe it's safer to have these assertions here in case when we attempt to remove those assumptions.